### PR TITLE
rust-client: Add support for QUIC transport

### LINKED ISSUE #41: rust-client: Add QUIC transport
_(Blocked until a new version of libp2p with libp2p-quic is released.)_

`rust-libp2p` recently merged an experimental QUIC transport (libp2p/rust-libp2p#2883). 
This QUIC implementation currently does not fully support dcutr in the sense that `Transport::dial_as_listener` is not support, which is needed if the local peer initiates a dcutr upgrade.
However, as a punchr client we dial remote peers via the relay, and the **remote** initiates the dcutr upgrade. This means that our local peer does a normal `Transport::dial`. Therefore we could already add QUIC to the rust-punchr client.
I did a short quic(k) test locally and got quite a few successful hole-punches via QUIC.

### PR DESCRIPTION
**Pull request by elenaf9**
Add support for the QUIC transport to the rust-client.
Resolves #41.
Since the IPFS network does not support the new `/quic-v1` codepoint yet we have to use the old `/quic` one, interpreted as draft-29. This is not supported by `rust-libp2p` `v0.50.0` / `libp2p-quic` `v0.7.0-alpha`, thus this PR uses right now libp2p from libp2p/rust-libp2p#3151.

### CODE REVIEW

File: rust-client/Cargo.toml
Comment (mxinden) [inline rust-client/Cargo.toml:14]: As discussed in a call, let's use a commit hash here.

File: rust-client/src/main.rs
Comment (mxinden) [inline rust-client/src/main.rs:214]: This would not allow a DNS address for QUIC, right? E.g. `/dns6/max-inden.de/udp/1234/quic`.

I think the easiest fix would be to wrap the `quic_transport` with another `DnsConfig::system`.

**Review by mxinden — COMMENTED** [review]
Thanks for preparing this!

File: rust-client/src/main.rs
Comment (elenaf9) [reply rust-client/src/main.rs]: Done in 780e104.

File: rust-client/Cargo.toml
Comment (elenaf9) [reply rust-client/Cargo.toml]: :+1: 63c45cf

**Comment by elenaf9** [comment]
@MarcoPolo nix build in the CI fails, I think because we are now using `libp2p` from git. Any idea how to fix this?

**Comment by elenaf9** [comment]
@mxinden I would already merge this PR. Once libp2p/rust-libp2p#3151 is merged we can change the git rev to the merge commit or wait for the patch release of `libp2p-quic`.

**Review by mxinden — APPROVED** [review]
Merging here sounds good to me.

**Comment by mxinden** [comment]
//CC @MarcoPolo we broke the Nix build, given that the Rust client now depends on a git reference. We prioritized QUIC over Nix. Please speak up in case that is a big issue.

**Comment by MarcoPolo** [comment]
Not a problem, looks like an easy fix. I'll have something up in a bit

**Pull request merged**
